### PR TITLE
[flink] Fixed the issue where deletes could not be performed when the table is in dynamic bucket mode.

### LINKED ISSUE #2956: [Bug] delete sql will fail when creating a table without explicitly declaring the number of buckets.
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Paimon version

0.7

### Compute Engine

flink-1.17.2

### Minimal reproduce step

```
CREATE TABLE  if not EXISTS use_be_hours_4 (
    user_id BIGINT,
    item_id BIGINT,
    behavior STRING,
    dt STRING,
    hh STRING,
    PRIMARY KEY (user_id) NOT ENFORCED
) 
with (
  'manifest.format'='orc',
  'changelog-producer' = 'lookup',
  'incremental-between-scan-mode'='changelog',
  'changelog-producer.row-deduplicate'='true'
  
);


INSERT INTO use_be_hours_4
VALUES
    (1, 1, 'watch', '2022-01-01', '10'),
    (2, 2, 'like', '2022-01-01', '10'),
     (0, 0, 'watch', '2022-01-01', '10'); 

-- ERROR HAPPEN AS BELOW     
delete from use_be_hours_4 where user_id =2;
```

ERROR MSG:
```
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Can't extract bucket from row in dynamic bucket mode, you should use 'TableWrite.write(InternalRow row, int bucket)' method.
	at org.apache.paimon.table.TableUtils.deleteWhere(TableUtils.java:74)
	at org.apache.paimon.flink.sink.SupportsRowLevelOperationFlinkTableSink.executeDeletion(SupportsRowLevelOperationFlinkTableSink.java:175)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:889)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:874)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:991)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:765)
	at org.dinky.executor.DefaultTableEnvironment.executeSql(DefaultTableEnvironment.java:300)
	at org.dinky.executor.Executor.executeSql(Executor.java:208)
	at org.dinky.job.builder.JobDDLBuilder.run(JobDDLBuilder.java:47)
	at org.dinky.job.JobManager.executeSql(JobManager.java:339)
	... 136 more
Caused by: java.lang.IllegalArgumentException: Can't extract bucket from row in dynamic bucket mode, you should use 'TableWrite.write(InternalRow row, int bucket)' method.
	at org.apache.paimon.table.sink.DynamicBucketRowKeyExtractor.bucket(DynamicBucketRowKeyExtractor.java:44)
	at org.apache.paimon.table.sink.TableWriteImpl.toSinkRecord(TableWriteImpl.java:148)
	at org.apache.paimon.table.sink.TableWriteImpl.writeAndReturn(TableWriteImpl.java:125)
	at org.apache.paimon.table.sink.TableWriteImpl.write(TableWriteImpl.java:116)
	at org.apache.paimon.table.TableUtils.deleteWhere(TableUtils.java:67)
	... 145 more
```

However, if I explicitly declare the number of buckets when creating the table as in the statement below, the issue does not occur.

```
CREATE TABLE  if not EXISTS use_be_hours_4 (
    user_id BIGINT,
    item_id BIGINT,
    behavior STRING,
    dt STRING,
    hh STRING,
    PRIMARY KEY (user_id) NOT ENFORCED
) 
with (
  'manifest.format'='orc',
  'changelog-producer' = 'lookup',
  'incremental-between-scan-mode'='changelog',
  'changelog-producer.row-deduplicate'='true',
  'bucket' = '1'
  
);
```

### What doesn't meet your expectations?

I can run successfully regardless of whether I explicitly declare the number of buckets when creating the table (as it defaults to one bucket if not declared).

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by hzjhjjyy**
### Purpose

Linked issue: close #2956 

and
1, add validation for bucket configuration.
2, fix typo.

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java [hunk 1]
```diff
@@ -233,8 +233,9 @@ default boolean writeFileUtf8(Path path, String content) throws IOException {
         Path tmp = path.createTempPath();
         boolean success = false;
         try {
-            try (PositionOutputStream out = newOutputStream(tmp, false)) {
-                OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+            try (PositionOutputStream out = newOutputStream(tmp, false);
+                    OutputStreamWriter writer =
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java:237]: Ditto.

File: paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java [hunk 2]
```diff
@@ -207,9 +207,9 @@ default void checkOrMkdirs(Path path) throws IOException {
 
     /** Read file to UTF_8 decoding. */
     default String readFileUtf8(Path path) throws IOException {
-        try (SeekableInputStream in = newInputStream(path)) {
-            BufferedReader reader =
-                    new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
+        try (SeekableInputStream in = newInputStream(path);
+                BufferedReader reader =
```
[the remark this answers is not in the document]
Comment (hzjhjjyy) [reply paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java:211]: It’s to add the remaining readers/writers to the try-with-resources block.

[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java:211]: It is just a wrapper, we already close the underly stream.

[the remark this answers is not in the document]
Comment (hzjhjjyy) [reply paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java]: ok, removed.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**